### PR TITLE
PAG variant for HunyuanDiT, PAG refactor

### PR DESCRIPTION
**Pull request by a-r-r-o-w**
# What does this PR do?

Part of #8785.

<details>
<summary> Code </summary>

```python
import torch
from diffusers import AutoPipelineForText2Image

pipe = AutoPipelineForText2Image.from_pretrained(
    "Tencent-Hunyuan/HunyuanDiT-v1.2-Diffusers",
    torch_dtype=torch.float16,
    enable_pag=True,
).to("cuda")

prompt = "A whimsical and creative image depicting a hybrid creature that is a mix of a waffle and a hippopotamus. This imaginative creature features the distinctive, bulky body of a hippo, but with a texture and appearance resembling a golden-brown, crispy waffle. The creature might have elements like waffle squares across its skin and a syrup-like sheen. It’s set in a surreal environment that playfully combines a natural water habitat of a hippo with elements of a breakfast table setting, possibly including oversized utensils or plates in the background. The image should evoke a sense of playful absurdity and culinary fantasy."
chinese_prompt = "一幅异想天开且富有创意的图像，描绘了华夫饼和河马的混合生物。这种富有想象力的生物具有河马独特而庞大的身体，但其质地和外观类似于金棕色的脆华夫饼。这种生物的皮肤上可能有华夫饼方块之类的元素，并且有糖浆般的光泽。它坐落在一个超现实的环境中，有趣地将河马的天然水生栖息地与早餐桌设置的元素结合在一起，背景可能包括超大的餐具或盘子。图像应该唤起一种俏皮的荒诞和烹饪幻想的感觉。"

for layer_indices in [[], [16, 17, 18, 19], [20, 21, 22, 23], [16, 17, 20, 23, 24], [14, 15, 20, 21, 22, 23]]:
    layer_str = "-".join(str(x) for x in layer_indices)
    pipe.set_pag_applied_layers(layer_indices)
    print(f"Processing {layer_str}")

    image = pipe(
        prompt=chinese_prompt,
        num_inference_steps=25,
        guidance_scale=4,
        pag_scale=2,
    ).images[0]
    image.save(f"hunyuandit_{layer_str}.png")
```
</details>

<table align="center">
<tr>
  <td align="center" colspan=2>w/o PAG</td>
</tr>
<tr>
  <td align="center" colspan=2></td>
</tr>
<tr>
  <td align="center"><b>[16, 17, 18, 19] PAG 2.0</b></td>
  <td align="center"><b>[20, 21, 22, 23] PAG 2.0</b></td>
</tr>
<tr>
  <td align="center"></td>
  <td align="center"></td>
</tr>
<tr>
  <td align="center"><b>[16, 17, 20, 23, 24] PAG 2.0</b></td>
  <td align="center"><b>[14, 15, 20, 21, 22, 23] PAG 2.0</b></td>
</tr>
<tr>
  <td align="center"></td>
  <td align="center"></td>
</tr>
</table>

TLDR; Layers 16-28 seem to be the sweet spot for applying PAG in Hunyuan. High CFG causes final image to have many artifacts, so values between 2-5 seem to work best. Low PAG scale between 1-3 also yields better results. Higher PAG and CFG causes the hilt of the sword to become more curved and causes artifacts near the hand, nose and whiskers, and it looks more blown-out and high constrast-y.

<details>
<summary> Ablations </summary>

- **Prompt**: A Puss in Boots, armed with a bright silver sword and clad in armor, stood on top of a pile of gold coins. The background was a dark-colored cave, and the image was dotted with golden coins.
- **Top-to-Bottom**: PAG Scale `[0, 0.5, 2.5, 4]`
- **Left-to-Right**: Guidance Scale `[1.0, 2.5, 5.0, 7.0]`

<table align="center">
<tr>
  <td align="center"><b> layers 4-5-6-7 </b></td>
</tr>
<tr>
  <td align="center"></td>
</tr>

<tr>
  <td align="center"><b> layers 12-13-14-15 </b></td>
</tr>
<tr>
  <td align="center"></td>
</tr>

<tr>
  <td align="center"><b> layers 10-11-12-13-14 </b></td>
</tr>
<tr>
  <td align="center"></td>
</tr>

<tr>
  <td align="center"><b> layers 16-17-18-19 </b></td>
</tr>
<tr>
  <td align="center"></td>
</tr>

<tr>
  <td align="center"><b> layers 20-21-22-23 </b></td>
</tr>
<tr>
  <td align="center"></td>
</tr>

<tr>
  <td align="center"><b> layers 20-21-22-23-24 </b></td>
</tr>
<tr>
  <td align="center"></td>
</tr>

<tr>
  <td align="center"><b> layers 24-25-26-27 </b></td>
</tr>
<tr>
  <td align="center"></td>
</tr>
</table>
</details>

TODO:
- [ ] Experiment with masked attention processor variant

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@yiyixuxu @asomoza @sunovivid

### CODE REVIEW

File: src/diffusers/pipelines/pag/pag_utils.py
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/pag/pag_utils.py:260]:
```suggestion
            if proc.__class__ in (PAGCFGIdentitySelfAttnProcessor2_0, PAGIdentitySelfAttnProcessor2_0):
```

**Comment by yiyixuxu** [comment]
ahh cool! it looks like PAG improved a lot, no?! 
it does not make a river or some sort of water in the background though, but I think it's probably the limitation of hunyuan 

is the Chinese prompt still work better than English for hunyuan-dit 1.2?

File: src/diffusers/pipelines/pag/pag_utils.py
Comment (yiyixuxu) [inline src/diffusers/pipelines/pag/pag_utils.py:265]: I know it is pretty natural to inherit from `PAGMixin` ( PAGMixin was not very designed to extend to other models I think) 
Since we really want to keep the level of abstraction minimum in diffusers so let's try to come up with ideas to not do make `HunyuanDiTPAGMixin` inherit from `PAGMixin` ( @sayakpaul is refactoring that on pixart pag PR https://github.com/huggingface/diffusers/pull/8921/files#r1689002227, I think we only need one PAG MIxin for all transformer models, so please share ideas there if you have any!)

**Comment by asomoza** [comment]
This one is more similar to SD3 in that each layer (except 3) changes more the style than the details. It still makes the image a lot better and depending on the layer(s) you can get different styles.

With this one, it fixes the hands and add more details to the clothes and other parts.

|W/O PAG|PAG|
|---|---|
|||

Other interesting effects when applying PAG to some layers are that for the 2nd it makes a "drawing" style image and with the 3rd it takes away every detail leaving mainly the composition.

|layer 2|layer 3|
|---|---|
|20240723230256_2013467037_layer_2|20240723230324_2013467037_layer_3|

The other interesting ones are:

|9|14|21|
|---|---|---|
|20240723230613_2013467037_layer_9|20240723230836_2013467037_layer_14|20240723231157_2013467037_layer_21|

As all the other transformers models with PAG, it's helps with the generation and they're cool to play with.

**Comment by yiyixuxu** [comment]
@asomoza super cool analysis :)

**Comment by a-r-r-o-w** [comment]
@yiyixuxu WDYT about the implementation? I've tried to make it so that handling both unet or transformer models would be easy, without too many code branches to check `blocks`, `transformer_blocks`, `down_blocks`, etc. using PEFT inspired regex matching.
- I really like the PEFT method and think this would keep things simple to use. You can still do things like `mid`, `down_blocks.0` and so on.
- You can set your own custom attention processors when calling `set_pag_enabled_layers`. This will make it easy for PixartPAG or SD3PAG to directly use PAGMixin by setting `pag_attention_processors` in `__init__()` like done here for Hunyuan
- Seems to work well for both unet/transformer case and reduces many LOC without needing to inherit from the PAGMixin for per-model case (Pixart, Hunyuan, SD3, etc.) You only need to know what the blocks are called. If you don't, you can simply do `.*blocks.*(idx_1|idx_2|...).*` because most of our internal modeling end are name `{something}_blocks`
- I can't recall if we have any layer naming for self-attention other than `attn1` so I've added it as a parameter. LMK if we don't need it.

Once it's good to go, I can update all tests once the other open PAG variants are merged, or their respective authors can adopt the mixin implementation from here. It is a bit backward breaking, but I think it's okay since we haven't released a Diffusers version with PAG yet so some unstability could be expected.

<details>
<summary> Here's some code for demo </summary>

```python
import torch
from diffusers import AutoPipelineForText2Image

pipe = AutoPipelineForText2Image.from_pretrained(
    "Tencent-Hunyuan/HunyuanDiT-v1.2-Diffusers",
    torch_dtype=torch.float16,
    enable_pag=True,
).to("cuda:0")

prompt = "A whimsical and creative image depicting a hybrid creature that is a mix of a waffle and a hippopotamus. This imaginative creature features the distinctive, bulky body of a hippo, but with a texture and appearance resembling a golden-brown, crispy waffle. The creature might have elements like waffle squares across its skin and a syrup-like sheen. It’s set in a surreal environment that playfully combines a natural water habitat of a hippo with elements of a breakfast table setting, possibly including oversized utensils or plates in the background. The image should evoke a sense of playful absurdity and culinary fantasy."
chinese_prompt = "一幅异想天开且富有创意的图像，描绘了华夫饼和河马的混合生物。这种富有想象力的生物具有河马独特而庞大的身体，但其质地和外观类似于金棕色的脆华夫饼。这种生物的皮肤上可能有华夫饼方块之类的元素，并且有糖浆般的光泽。它坐落在一个超现实的环境中，有趣地将河马的天然水生栖息地与早餐桌设置的元素结合在一起，背景可能包括超大的餐具或盘子。图像应该唤起一种俏皮的荒诞和烹饪幻想的感觉。"

test_layers = [
    [],
    "blocks.4",
    "blocks.10",
    "blocks.16",
    "blocks.20",
    "blocks\.22",
    r"blocks.24",
    "blocks.(16|17|18|19)",
    r"blocks\.(20|21|22|23)",
    ["blocks.16", "blocks.17", "blocks\.(20|23)", r"blocks.24"],
]

for index, layers in enumerate(test_layers):
    pipe.set_pag_applied_layers(layers)
    print(f"Processing {layers}")

    image = pipe(
        prompt=prompt,
        num_inference_steps=25,
        guidance_scale=4,
        pag_scale=2,
    ).images[0]
    image.save(f"hunyuandit_{index}_eng.png")

    image = pipe(
        prompt=chinese_prompt,
        num_inference_steps=25,
        guidance_scale=4,
        pag_scale=2,
    ).images[0]
    image.save(f"hunyuandit_{index}_chn.png")
```
</details>

> ahh cool! it looks like PAG improved a lot, no?!
> it does not make a river or some sort of water in the background though, but I think it's probably the limitation of hunyuan

Yes, seems like a limitation with Hunyuan. Never got it to work for really descriptive prompts with all expected objects/styles.

> is the Chinese prompt still work better than English for hunyuan-dit 1.2?

Have not really dived too deep into this. I assumed it would be better for chinese because of model origin

cc @sayakpaul @asomoza

File: src/diffusers/pipelines/pag/pipeline_pag_hunyuandit.py
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/pag/pipeline_pag_hunyuandit.py:248]: For Pixart/SD3, deriving from PAGMixin and changing this would be all the changes needed I think. I have also enforced pag_applied_layers being either a string or list of strings, and disallowed integers to be passed for layers.

File: src/diffusers/pipelines/pag/pag_utils.py
Comment (yiyixuxu) [inline src/diffusers/pipelines/pag/pag_utils.py:42]: I think it's better to raise an error here

File: src/diffusers/pipelines/pag/pag_utils.py
Comment (yiyixuxu) [inline src/diffusers/pipelines/pag/pag_utils.py:152]:
```suggestion
        pag_attn_processors: Optional[Tuple[AttentionProcessor, AttentionProcessor]] = ((PAGCFGIdentitySelfAttnProcessor2_0(), PAGIdentitySelfAttnProcessor2_0())
```
let's just

File: src/diffusers/pipelines/pag/pag_utils.py
Comment (yiyixuxu) [inline src/diffusers/pipelines/pag/pag_utils.py:217]: let's just make sure it always has `_pag_attn_processor` set to simplify things

File: src/diffusers/pipelines/pag/pag_utils.py
Comment (yiyixuxu) [inline src/diffusers/pipelines/pag/pag_utils.py:31]: where did this check go?

**Review by yiyixuxu — COMMENTED** [review]
thanks! very nice!
I left some comments, also would wait for this to merge and refactor a little bit more https://github.com/huggingface/diffusers/pull/8921/files#diff-f28d850df75fd10c4100a8de8d944f3bbab67e03fd9d3122fb7cce3572a5eebc

File: src/diffusers/pipelines/pag/pag_utils.py
Comment (a-r-r-o-w) [reply src/diffusers/pipelines/pag/pag_utils.py:31]: why do we need this check any more btw? since the matching is now regex based, a user of the library could do a wide variety of things that would be near to impossible to validate. how could we verify this?

**Comment by a-r-r-o-w** [comment]
@yiyixuxu thanks for the review! i've addressed them for the new design. let me know if there's anything else that i missed

cc @sayakpaul for viz. we won't be able to use integer layer indexes any more to apply PAG (let's say for PixArt). i've made it so that we use strings throughout. with the new design, one can apply PAG in many ways (due to regex matching) and the intended ways to use it is by specifying `blocks.{layer}` for single layers, or `blocks.({layer_1}|{layer_2}|...)` etc. please review the code example above

let's try and get this merged soon since we can't introduce backward breaking changes after next release

**Comment by sayakpaul** [comment]
block.{layer_idx} works for me. We need to get that in the release otherwise we would introduce something that we know will break after this PR is merged. 

@yiyixuxu let’s maybe try to also get the SD3 PAG PR done as well?

**Review by yiyixuxu — APPROVED** [review]
very very nice!

**Comment by sayakpaul** [comment]
@a-r-r-o-w could we fix the PixArt Sigma tests as well? Can merge afterward.

File: src/diffusers/pipelines/pag/pipeline_pag_pixart_sigma.py
Comment (sayakpaul) [inline src/diffusers/pipelines/pag/pipeline_pag_pixart_sigma.py:135]: We need to update the example doc too and the default value here:

https://github.com/huggingface/diffusers/blob/ebf3ab1477dd480df1b8dd5d97a7b4aa3822716b/src/diffusers/pipelines/pag/pipeline_pag_pixart_sigma.py#L64

https://github.com/huggingface/diffusers/blob/ebf3ab1477dd480df1b8dd5d97a7b4aa3822716b/src/diffusers/pipelines/pag/pipeline_pag_pixart_sigma.py#L167

File: src/diffusers/pipelines/pag/pag_utils.py
Comment (sayakpaul) [reply src/diffusers/pipelines/pag/pag_utils.py:31]: IIUC, the regex itself should error out then right if there's no matching pattern? If we're not doing it already, could we add one?

File: src/diffusers/pipelines/pag/pag_utils.py
Comment (sayakpaul) [inline src/diffusers/pipelines/pag/pag_utils.py:73]: ❤️

File: src/diffusers/pipelines/pag/pag_utils.py
Comment (sayakpaul) [inline src/diffusers/pipelines/pag/pag_utils.py:79]: ❤️

File: src/diffusers/pipelines/pag/pag_utils.py
Comment (sayakpaul) [inline src/diffusers/pipelines/pag/pag_utils.py:218]: Curious on the choice of `tuple` as a data structure here. Why not a set in case performance is concerned.

File: src/diffusers/pipelines/pag/pag_utils.py
Comment (sayakpaul) [inline src/diffusers/pipelines/pag/pag_utils.py:151]: There should Hunyuan processors as well, right?

File: tests/pipelines/pag/test_pag_pixart_sigma.py
Comment (sayakpaul) [inline tests/pipelines/pag/test_pag_pixart_sigma.py:130]: Shouldn't this be "block.1"?

**Review by sayakpaul — COMMENTED** [review]
LGTM, thank you. The refactor is beautiful. My comments are mostly minor. LMK in case something isn't clear enough. 

Let's fix the PixArt tests and merge.

File: src/diffusers/pipelines/pag/pag_utils.py
Comment (a-r-r-o-w) [reply src/diffusers/pipelines/pag/pag_utils.py:218]: Ah yes, apologies for the oversight. Should be set

File: tests/pipelines/pag/test_pag_pixart_sigma.py
Comment (a-r-r-o-w) [reply tests/pipelines/pag/test_pag_pixart_sigma.py:130]: Yes, my bad. Will update these. The intended usage is "blocks.{index}" or some regex combo of layers, but just strings like this work too (in this case) since the PAG logic is able to fulfil the 3 conditions mentioned. It will give unexpected behaviour if you have layers named like:

```
something.blocks.1.another_something.1.attn1
something.blocks.42.another_something.1.attn1
```

In this case, the intent was probably to only apply to `something.blocks.1` but PAG will get applied to both. But the general guide would be to follow `blocks.{}` convention because we use it consistently in naming our blocks. With regex, there are many ways to shoot yourself in the foot and so we can't really account for all cases if users don't want to follow the general guide.

File: src/diffusers/pipelines/pag/pag_utils.py
Comment (a-r-r-o-w) [reply src/diffusers/pipelines/pag/pag_utils.py:151]: Yes, correct. Will update

Comment (a-r-r-o-w) [reply src/diffusers/pipelines/pag/pag_utils.py:151]: Actually, no. If this was a type hint, we would have had Hunyuan ones as well. I am instantiating with default pag attention processors here if the user does not provide any (which is how we do it in most pag pipelines). For Hunyuan, we explicitly pass the pag_attn_processors parameter.

File: tests/pipelines/pag/test_pag_pixart_sigma.py
Comment (sayakpaul) [reply tests/pipelines/pag/test_pag_pixart_sigma.py:130]: Yeah fair enough. Maybe we can update the documentation to be explicit about that? PAG is a very widely used feature, so, need to be rigorous. WDYT?

File: src/diffusers/pipelines/pag/pag_utils.py
Comment (a-r-r-o-w) [reply src/diffusers/pipelines/pag/pag_utils.py:31]: We do raise an error if no matches are found via regex, so that should be covered

**Comment by a-r-r-o-w** [comment]
@sayakpaul Why does it say that it can't find the `pag_attn_processors` property? The only difference between Hunyuan/Pixart and other PAG pipelines is that these are the only two that have a single mixin, whreas the remaining pipelines have multiple mixins. I looked at the MRO's and PAGMixin clearly exists so not sure why it can't access these.

image


For expected slices, will fix them in a bit

**Comment by sayakpaul** [comment]
> Why does it say that it can't find the pag_attn_processors property? 

I would suggest taking a look more closely. If that does not cut it I can take a closer look. 

Why do we have to update the slices? Usually, that means there might be some incorrect codepath being taken.

**Comment by a-r-r-o-w** [comment]
> Why do we have to update the slices? Usually, that means there might be some incorrect codepath being taken.

The reason slices need to be updated in AnimateDiffPAG is because they were created for original PAG implementation where if you specified "mid", only `attention.attn1` and `motion_module.attn1` got activated. With the current implementation, when you specify "mid", the layers activated are `attention.attn1`, `motion_module.attn1`, `motion_module.attn2`. This is because:
- both attn1 and attn2 of motion module is self attention
- in previous implementation, we were checking for "attn1" specifically, but as mentioned by YiYi [here](https://github.com/huggingface/diffusers/pull/8936#discussion_r1700970344), I thought it's best to use `is_cross_attention` flag as well to determine whether a layer is self-attention or not.

Now that attn2 is also used by default (quality-wise it is similar), we'd need to update slices.

**Comment by a-r-r-o-w** [comment]
@yiyixuxu @sayakpaul For some reason, MRO is fighting me here and I'm not completely sure why.

<details>
<summary> Dump </summary>

```python
>>> from diffusers import HunyuanDiTPAGPipeline
>>> pipe = HunyuanDiTPAGPipeline(None, None, None, None, None)
You have disabled the safety checker for <class 'diffusers.pipelines.pag.pipeline_pag_hunyuandit.HunyuanDiTPAGPipeline'> by passing `safety_checker=None`. Ensure that you abide to the conditions of the Stable Diffusion license and do not expose unfiltered results in services or applications open to the public. Both the diffusers team and Hugging Face strongly recommend to keep the safety filter enabled in all public facing circumstances, disabling it only for use-cases that involve analyzing network behavior or auditing its results. For more information, please have a look at https://github.com/huggingface/diffusers/pull/254 .
>>> dir(pipe)
['__call__', '__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattr__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_apply_perturbed_attention_guidance', '_callback_tensor_inputs', '_deprecated_kwargs', '_dict_from_json_file', '_exclude_from_cpu_offload', '_execution_device', '_get_init_keys', '_get_pag_scale', '_get_signature_keys', '_get_signature_types', '_internal_dict', '_is_onnx', '_load_connected_pipes', '_optional_components', '_pag_attn_processors', '_prepare_perturbed_attention_guidance', '_set_pag_attn_processor', '_upload_folder', 'check_inputs', 'components', 'config', 'config_name', 'default_sample_size', 'device', 'disable_attention_slicing', 'disable_xformers_memory_efficient_attention', 'do_classifier_free_guidance', 'do_pag_adaptive_scaling', 'do_perturbed_attention_guidance', 'download', 'dtype', 'enable_attention_slicing', 'enable_model_cpu_offload', 'enable_sequential_cpu_offload', 'enable_xformers_memory_efficient_attention', 'encode_prompt', 'extract_init_dict', 'feature_extractor', 'from_config', 'from_pipe', 'from_pretrained', 'get_config_dict', 'guidance_rescale', 'guidance_scale', 'has_compatibles', 'hf_device_map', 'ignore_for_config', 'image_processor', 'interrupt', 'load_config', 'maybe_free_model_hooks', 'model_cpu_offload_seq', 'name_or_path', 'num_timesteps', 'numpy_to_pil', 'pag_adaptive_scale', 'pag_applied_layers', 'pag_attn_processors', 'pag_scale', 'prepare_extra_step_kwargs', 'prepare_latents', 'progress_bar', 'push_to_hub', 'register_modules', 'register_to_config', 'remove_all_hooks', 'reset_device_map', 'run_safety_checker', 'safety_checker', 'save_config', 'save_pretrained', 'scheduler', 'set_attention_slice', 'set_pag_applied_layers', 'set_progress_bar_config', 'set_use_memory_efficient_attention_xformers', 'text_encoder', 'text_encoder_2', 'to', 'to_json_file', 'to_json_string', 'tokenizer', 'tokenizer_2', 'transformer', 'vae', 'vae_scale_factor']
>>> pipe.pag_attn_processors
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/aryan/work/diffusers/src/diffusers/configuration_utils.py", line 143, in __getattr__
    raise AttributeError(f"'{type(self).__name__}' object has no attribute '{name}'")
AttributeError: 'HunyuanDiTPAGPipeline' object has no attribute 'pag_attn_processors'. Did you mean: '_pag_attn_processors'?
>>> pipe.pag_scale
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/aryan/work/diffusers/src/diffusers/configuration_utils.py", line 143, in __getattr__
    raise AttributeError(f"'{type(self).__name__}' object has no attribute '{name}'")
AttributeError: 'HunyuanDiTPAGPipeline' object has no attribute 'pag_scale'
```
</details>

Also: https://github.com/huggingface/diffusers/blob/2b7609961034d9b785c3836a169cb18e6ccb0a39/src/diffusers/configuration_utils.py#L136

If you log the `__dict__` here, none of the `@property` methods show up for Hunyuan or Pixart (not even things like `do_classifier_free_guidance` or `guidance_scale`). However, for all other pipelines it does show up so I'm not sure what we are missing here for Hunyuan and Pixart. The only difference is that the latter two inherit from DiffusionPipeline and PAGMixin whereas the others inherit from multiple different mixins.

**Comment by sayakpaul** [comment]
> Edit: the easier fix would be to just specify attn1 layers to stay consistent with the test and example code here

That is indeed better. 

Regarding the MRO, I am taking a look. Will update here if I find anything.

**Comment by sayakpaul** [comment]
@a-r-r-o-w I think I found the bug. Here's a patch:

[fix_pag_refactor.patch](https://github.com/user-attachments/files/16492071/fix_pag_refactor.patch)

Let me know if this resolves your issue.

**Comment by sayakpaul** [comment]
Since we have Yiyi's approval, I think we can merge this.

**Comment by a-r-r-o-w** [comment]
Thanks for all the reviews on this one @sayakpaul @yiyixuxu and the design iterations, @asomoza for the really cool examples, and ofcourse @sunovivid and team for the awesome technique!

Will merge once CI is green

**Pull request merged**
